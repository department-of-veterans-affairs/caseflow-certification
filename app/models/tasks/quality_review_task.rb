# frozen_string_literal: true

##
# Task to track when an appeal has been randomly selected to be quality reviewed by the Quality Review team.

class QualityReviewTask < Task
  before_create :prevent_quality_review_task_creation_for_unrecognized_appellants

  scope :created_this_month, -> { where(created_at: Time.zone.now.beginning_of_month..Time.zone.now.end_of_month) }

  def available_actions(user)
    return super if assigned_to != user

    [
      Constants.TASK_ACTIONS.MARK_COMPLETE.to_h,
      Constants.TASK_ACTIONS.QR_RETURN_TO_JUDGE.to_h,
      Constants.TASK_ACTIONS.TOGGLE_TIMED_HOLD.to_h,
      Constants.TASK_ACTIONS.CANCEL_TASK.to_h
    ]
  end

  def self.create_from_root_task(root_task)
    create!(assigned_to: QualityReview.singleton, parent_id: root_task.id, appeal: root_task.appeal)
  end

  def update_parent_status
    # QualityReviewTasks may be assigned to organizations or individuals. However, for each appeal that goes through
    # quality review a task assigned to the organization will exist (even if there is none assigned to an
    # individual). To prevent creating duplicate BvaDispatchTasks only create one for the organization task.
    BvaDispatchTask.create_from_root_task(root_task) if assigned_to == QualityReview.singleton
    super
  end

  private

  def prevent_quality_review_task_creation_for_unrecognized_appellants
    unless FeatureToggle.enabled?(:allow_unrecognized_appellant_dispatch)
      fail NotImplementedError if appeal.claimant.is_a?(OtherClaimant)
    end
  end
end
