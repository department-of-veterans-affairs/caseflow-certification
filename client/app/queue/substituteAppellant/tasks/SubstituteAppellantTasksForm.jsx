import React from 'react';
import PropTypes from 'prop-types';
import { FormProvider, useFieldArray, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { css } from 'glamor';

import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import { AddAdminTaskForm } from 'app/queue/colocatedTasks/AddAdminTaskForm/AddAdminTaskForm';
import {
  SUBSTITUTE_APPELLANT_CREATE_TASKS_TITLE,
  SUBSTITUTE_APPELLANT_SELECT_APPELLANT_SUBHEAD,
} from 'app/../COPY';
import CheckoutButtons from 'app/queue/docketSwitch/grant/CheckoutButtons';
import Button from 'app/components/Button';
import { KeyDetails } from './KeyDetails';

const schema = yup.object().shape({
  newTasks: yup.array(
    yup.object().shape({
      type: yup.string().required(),
      instructions: yup.string().required(),
    })
  ),
});

const sectionStyle = css({ marginBottom: '24px' });

export const SubstituteAppellantTasksForm = ({
  appealId,
  existingValues,
  nodDate,
  dateOfDeath,
  substitutionDate,
  onBack,
  onCancel,
  onSubmit,
}) => {
  const methods = useForm({
    // Use this for repopulating form from redux when user navigates back
    defaultValues: {
      ...existingValues,
      newTasks: existingValues?.newTasks ?? [],
    },
    resolver: yupResolver(schema),
  });
  const { control, handleSubmit } = methods;

  const { fields, append, remove } = useFieldArray({
    control,
    name: 'newTasks',
  });

  return (
    <FormProvider {...methods}>
      <form onSubmit={handleSubmit(onSubmit)}>
        <AppSegment filledBackground>
          <h1>{SUBSTITUTE_APPELLANT_CREATE_TASKS_TITLE}</h1>
          <div {...sectionStyle}>
            {SUBSTITUTE_APPELLANT_SELECT_APPELLANT_SUBHEAD}
          </div>
          <KeyDetails
            appealId={appealId}
            nodDate={nodDate}
            dateOfDeath={dateOfDeath}
            substitutionDate={substitutionDate}
          />

          <React.Fragment>
            <h3 {...css({ marginBottom: '0' })}>
              <br />
              <strong>Would you like to add any additional tasks?</strong>
              <br />
            </h3>
            <div>
              {fields.map((item, idx) => (
                <AddAdminTaskForm
                  key={item.id}
                  item={item}
                  baseName={`newTasks[${idx}]`}
                  onRemove={() => remove(idx)}
                />
              ))}
            </div>
            <Button
              willNeverBeLoading
              dangerStyling
              styling={css({ marginTop: '1rem' })}
              name="+ Add task"
              onClick={() => append({ type: null, instructions: '' })}
            />
          </React.Fragment>
        </AppSegment>
        <div className="controls cf-app-segment">
          <CheckoutButtons
            onCancel={onCancel}
            onBack={onBack}
            onSubmit={handleSubmit(onSubmit)}
            submitText="Continue"
          />
        </div>
      </form>
    </FormProvider>
  );
};
SubstituteAppellantTasksForm.propTypes = {
  appealId: PropTypes.string,
  existingValues: PropTypes.shape({}),
  nodDate: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string]),
  dateOfDeath: PropTypes.oneOfType([
    PropTypes.instanceOf(Date),
    PropTypes.string,
  ]),
  substitutionDate: PropTypes.oneOfType([
    PropTypes.instanceOf(Date),
    PropTypes.string,
  ]),
  onBack: PropTypes.func,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func,
};
