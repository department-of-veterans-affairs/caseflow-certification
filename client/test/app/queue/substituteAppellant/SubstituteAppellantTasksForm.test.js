import React from 'react';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import selectEvent from 'react-select-event';
import { axe } from 'jest-axe';
import { parseISO } from 'date-fns';

import { SubstituteAppellantTasksForm } from 'app/queue/substituteAppellant/tasks/SubstituteAppellantTasksForm';
import { MemoryRouter } from 'react-router';

describe('SubstituteAppellantTasksForm', () => {
  const onSubmit = jest.fn();
  const onCancel = jest.fn();

  const defaults = {
    appealId: 'abc123',
    nodDate: parseISO('2021-04-01'),
    dateOfDeath: parseISO('2021-04-15'),
    substitutionDate: parseISO('2021-04-20'),
    onCancel,
    onSubmit,
  };

  const setup = (props) =>
    render(
      <MemoryRouter>
        <SubstituteAppellantTasksForm {...defaults} {...props} />
      </MemoryRouter>
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('fires onCancel', async () => {
    setup();
    expect(onCancel).not.toHaveBeenCalled();

    await userEvent.click(screen.getByRole('button', { name: /cancel/i }));
    expect(onCancel).toHaveBeenCalled();
  });

  describe('with blank form', () => {
    it('renders default state correctly', () => {
      const { container } = setup();

      expect(container).toMatchSnapshot();
    });

    it('passes a11y testing', async () => {
      const { container } = setup();

      const results = await axe(container);

      expect(results).toHaveNoViolations();
    });

    // describe('form validation', () => {
    //   it('requires at least some tasks to have been entered', async () => {
    //     setup();

    //     const submit = screen.getByRole('button', { name: /continue/i });

    //     // Submit to trigger validation
    //     await userEvent.click(submit);

    //     expect(onSubmit).not.toHaveBeenCalled();
    //   });
    // });
  });

  describe('additional admin actions', () => {
    const clickAddTask = () =>
      userEvent.click(screen.getByRole('button', { name: /add task/i }));

    it('shows the form when button is pressed', async () => {
      setup();

      clickAddTask();

      expect(
        screen.getByRole('textbox', { name: /select the type of task/i })
      ).toBeInTheDocument();

      expect(
        screen.getByRole('textbox', { name: /instructions/i })
      ).toBeInTheDocument();

      expect(
        screen.getByRole('button', { name: /remove/i })
      ).toBeInTheDocument();

      // Still only one "Add Task" button
      expect(
        screen.queryAllByRole('button', { name: /add task/i }).length
      ).toBe(1);
    });

    it('supports adding multiple tasks', async () => {
      setup();

      clickAddTask();

      expect(
        screen.queryAllByRole('textbox', { name: /select the type of task/i }).
          length
      ).toBe(1);

      expect(
        screen.queryAllByRole('textbox', { name: /instructions/i }).length
      ).toBe(1);

      // Add a second task
      clickAddTask();

      expect(
        screen.queryAllByRole('textbox', { name: /select the type of task/i }).
          length
      ).toBe(2);

      expect(
        screen.queryAllByRole('textbox', { name: /instructions/i }).length
      ).toBe(2);

      expect(screen.queryAllByRole('button', { name: /remove/i }).length).toBe(
        2
      );

      // Still only one "Add Task" button
      expect(
        screen.queryAllByRole('button', { name: /add task/i }).length
      ).toBe(1);
    });

    it('submits added tasks', async () => {
      setup();

      clickAddTask();

      await selectEvent.select(
        screen.getByLabelText(/select the type of task/i),
        'IHP'
      );

      userEvent.type(
        screen.getByRole('textbox', { name: /instructions/i }),
        'foo bar'
      );

      await act(async () => {
        await userEvent.click(
          screen.getByRole('button', { name: /continue/i })
        );
      });

      expect(onSubmit.mock.calls[0][0]?.newTasks?.length).toBe(1);
      expect(onSubmit.mock.calls[0][0]?.newTasks?.[0]).toEqual(
        expect.objectContaining({
          type: 'IhpColocatedTask',
          instructions: 'foo bar',
        })
      );
    });
  });

  describe('with existing form data', () => {
    const existingValues = {
      // Insert task data here
    };

    it('renders default state correctly', () => {
      const { container } = setup({ existingValues });

      expect(container).toMatchSnapshot();
    });

    it('passes a11y testing', async () => {
      const { container } = setup({ existingValues });

      const results = await axe(container);

      expect(results).toHaveNoViolations();
    });

    describe('form validation', () => {
      it('fires onSubmit because everything is good to go', async () => {
        setup({ existingValues });

        const submit = screen.getByRole('button', { name: /continue/i });

        // Submit to trigger validation
        await userEvent.click(submit);

        await waitFor(() => {
          expect(onSubmit).toHaveBeenCalled();
        });
      });
    });
  });
});
