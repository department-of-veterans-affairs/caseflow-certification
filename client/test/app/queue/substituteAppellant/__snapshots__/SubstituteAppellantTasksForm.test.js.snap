// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubstituteAppellantTasksForm with blank form renders default state correctly 1`] = `
<div>
  <form>
    <section
      class="cf-app-segment cf-app-segment--alt"
      role="main"
    >
      <h1>
        Create task
      </h1>
      <div
        data-css-1sbm5ab=""
      >
        Complete the details below to create the appeal for the substitute for this case.
      </div>
      <section>
        <h2>
          Key details
        </h2>
        <ul
          class="css-a7zcn1"
        >
          <li>
            <strong>
              Notice of disagreement received
            </strong>
            <span>
              4/1/2021
            </span>
          </li>
          <li>
            <strong>
              Veteran date of death
            </strong>
            <span>
              4/15/2021
            </span>
          </li>
          <li>
            <strong>
              Substitution granted by the RO
            </strong>
            <span>
              4/20/2021
            </span>
          </li>
        </ul>
        <div
          class="css-2zljw9"
        >
          <a
            href="/queue/appeals/abc123"
            target="_blank"
          >
            View original case details 
            <i
              class="fa fa-external-link"
            />
          </a>
        </div>
      </section>
      <h3
        data-css-1m1s8ka=""
      >
        <br />
        <strong>
          Would you like to add any additional tasks?
        </strong>
        <br />
      </h3>
      <div />
      <button
        class="cf-submit usa-button-secondary"
        data-css-byo489=""
        id="button-+-Add-task"
        type="button"
      >
        + Add task
      </button>
    </section>
    <div
      class="controls cf-app-segment"
    >
      <div
        data-css-63oe3q=""
      >
        <span>
          <button
            class="cf-right-side usa-button"
            data-css-wn5k45=""
            id="button-next-button"
            type="button"
          >
            Continue
          </button>
        </span>
        <button
          class="cf-btn-link cf-right-side usa-button"
          id="button-cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`SubstituteAppellantTasksForm with existing form data renders default state correctly 1`] = `
<div>
  <form>
    <section
      class="cf-app-segment cf-app-segment--alt"
      role="main"
    >
      <h1>
        Create task
      </h1>
      <div
        data-css-1sbm5ab=""
      >
        Complete the details below to create the appeal for the substitute for this case.
      </div>
      <section>
        <h2>
          Key details
        </h2>
        <ul
          class="css-a7zcn1"
        >
          <li>
            <strong>
              Notice of disagreement received
            </strong>
            <span>
              4/1/2021
            </span>
          </li>
          <li>
            <strong>
              Veteran date of death
            </strong>
            <span>
              4/15/2021
            </span>
          </li>
          <li>
            <strong>
              Substitution granted by the RO
            </strong>
            <span>
              4/20/2021
            </span>
          </li>
        </ul>
        <div
          class="css-2zljw9"
        >
          <a
            href="/queue/appeals/abc123"
            target="_blank"
          >
            View original case details 
            <i
              class="fa fa-external-link"
            />
          </a>
        </div>
      </section>
      <h3
        data-css-1m1s8ka=""
      >
        <br />
        <strong>
          Would you like to add any additional tasks?
        </strong>
        <br />
      </h3>
      <div />
      <button
        class="cf-submit usa-button-secondary"
        data-css-byo489=""
        id="button-+-Add-task"
        type="button"
      >
        + Add task
      </button>
    </section>
    <div
      class="controls cf-app-segment"
    >
      <div
        data-css-63oe3q=""
      >
        <span>
          <button
            class="cf-right-side usa-button"
            data-css-wn5k45=""
            id="button-next-button"
            type="button"
          >
            Continue
          </button>
        </span>
        <button
          class="cf-btn-link cf-right-side usa-button"
          id="button-cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </form>
</div>
`;
